db/views/reporting_patient_visits_v03.sql

```diff
@@ -0,0 +1,102 @@
+-- Only most recent Encounter per patient per month. Encounters are ordered appropriately below.
+SELECT
+    DISTINCT ON (p.id, p.month_date)
+    p.id as patient_id,
+
+    p.month_date,
+    p.month,
+    p.quarter,
+    p.year,
+    p.month_string,
+    p.quarter_string,
+
+    p.assigned_facility_id AS assigned_facility_id,
+    p.registration_facility_id AS registration_facility_id,
+    p.recorded_at AT TIME ZONE 'UTC' AT TIME ZONE 'UTC' AS patient_recorded_at,
+
+    ------------------------------------------------------------
+    -- details of the visit: latest encounter, prescription drug and appointment
+
+    e.id AS encounter_id,
+    e.facility_id AS encounter_facility_id,
+    e.recorded_at AT TIME ZONE 'UTC' AT TIME ZONE 'UTC' AS encounter_recorded_at,
+
+    pd.id AS prescription_drug_id,
+    pd.facility_id AS prescription_drug_facility_id,
+    pd.recorded_at AT TIME ZONE 'UTC' AT TIME ZONE 'UTC' AS prescription_drug_recorded_at,
+
+    app.id AS appointment_id,
+    app.creation_facility_id AS appointment_creation_facility_id,
+    app.recorded_at AT TIME ZONE 'UTC' AT TIME ZONE 'UTC' AS appointment_recorded_at,
+
+    array_remove(
+        ARRAY[
+            (CASE WHEN to_char(e.recorded_at AT TIME ZONE 'utc' AT TIME ZONE (SELECT current_setting('TIMEZONE')), 'YYYY-MM') = month_string THEN e.facility_id END),
+            (CASE WHEN to_char(pd.recorded_at AT TIME ZONE 'utc' AT TIME ZONE (SELECT current_setting('TIMEZONE')), 'YYYY-MM') = month_string THEN pd.facility_id END),
+            (CASE WHEN to_char(app.recorded_at AT TIME ZONE 'utc' AT TIME ZONE (SELECT current_setting('TIMEZONE')), 'YYYY-MM') = month_string THEN app.creation_facility_id END)
+        ],
+        null
+    ) AS visited_facility_ids,
```

Comment:
I might have missed this column in the initial enhancements PR, my bad. This field seems a bit gnarly to me. Would it only contain facility IDs if the visit happened _this_ month? Eg. If I last visited in May, would the June "visited_facility_ids" be empty?

Intuitively, I'd think that I'd have a `months_since_visit`, and a `visit_facility_id` that is the facility of my most recent visit, regardless of how long ago.

Reply:
x-posting a comment from [elsewhere](https://simpledotorg.slack.com/archives/C01JJJM1NA1/p1626673495091300?thread_ts=1626431704.090700&cid=C01JJJM1NA1):

The follow up definitions in `patient_states` would be flawed regardless of the implementation of `visited_facility_ids`, so I’m in favour of removing this, and other follow up related columns as well _in the next PR_. I’ll know more after I deal with the one-off request to create follow up reports with the new definitions. 

- The current matviews hold only “_latest_” data in the month, which means if there are 2 BPs in 2 different facilities, we would only consider one of them in the reports, which is incorrect
- The views matview uses encounters, so it inherently suppresses either BP/Sugar reading to get us the latest information

My takeaway from all this is that “follow up” reports tell us about the facility’s activity/health, whereas the treatment-outcome etc tell us about the patient’s health. I.e, the follow up reports are not longitudinal patient reports, and we can’t use the patient-states table for it.

The current plan (in my head) is to create the follow up report elsewhere (a different matview perhaps) and add that information to the facility states matview.